[GlobalISel][AArch64] Tail call libcalls.

davemgreen opened the pull request:
This tries to allow libcalls to be tail called, using a similar method to DAG where the type is checked to make sure they match, and if so the backend, through lowerCall checks that the tailcall is valid for all arguments.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -565,9 +566,29 @@ static RTLIB::Libcall getRTLibDesc(unsigned Opcode, unsigned Size) {
+  GetReturnInfo(CC, F.getReturnType(), F.getAttributes(), OutsFn, TLI, DL);
+  SmallVector<ISD::OutputArg, 4> OutsLibcall;
+  GetReturnInfo(CC, Result.Ty, F.getAttributes(), OutsLibcall, TLI, DL);
+  return OutsFn.size() == OutsLibcall.size() &&
+         all_of(zip(OutsFn, OutsLibcall),
+                [](auto P) { return get<0>(P).VT == get<1>(P).VT; });
+}
```
Shouldn't this be using CCState::resultsCompatible? Exact type match is more conservative than necessary

davemgreen replied:
Ah - that is what I was after. I didn't know what it was called and hadn't looked in the right place for it. Thanks.

davemgreen replied:
This might be more difficult actually, and may need to defer to the target to get the correct info it needs.

davemgreen committed:
[GlobalISel][AArch64] Tail call libcalls.

This tries to allow libcalls to be tail called, using a similar method to DAG
where the type is checked to make sure they match, and if so the backend,
through lowerCall checks that the tailcall is valid for all arguments.

davemgreen commented:
I've changed this to work hopefully similar to how DAG works, where the return types are checked to makes sure they match and if so the target, through lowerCall checks the tailcall is valid for the arguments.

aemerson reviewed (approved):
LGTM, thanks!
